sympy/core/equation.py

```diff
@@ -0,0 +1,521 @@
+    >>> poly.applylhs(factor)
+    x*(a*x + b + c*x) = a*x**3 + b*x**3 + c*x
+    >>> poly.applylhs(collect,x)
+    b*x + x**2*(a + c) = a*x**3 + b*x**3 + c*x
+
+    ``.apply...`` also works with user defined python functions
+    >>> def addsquare(eqn):
+    ...     return eqn+eqn**2
+    ...
+    >>> t.apply(addsquare)
+    a**2 + a = b**2/c**2 + b/c
+    >>> t.applyrhs(addsquare)
+    a = b**2/c**2 + b/c
+    >>> t.apply(addsquare, side = 'rhs')
+    a = b**2/c**2 + b/c
+    >>> t.applylhs(addsquare)
+    a**2 + a = b/c
+    >>> addsquare(t)
+    a**2 + a = b**2/c**2 + b/c
+
+    Inaddition to ``.apply...`` there is also the less general ``.do``,
+    ``.dolhs``, ``.dorhs``, which only works for operations defined on the
+    ``Expr`` class (e.g.``.collect(), .factor(), .expand()``, etc...).
+    >>> poly.dolhs.collect(x)
+    b*x + x**2*(a + c) = a*x**3 + b*x**3 + c*x
+    >>> poly.dorhs.collect(x)
+    a*x**2 + b*x + c*x**2 = c*x + x**3*(a + b)
+    >>> poly.do.collect(x)
+    b*x + x**2*(a + c) = c*x + x**3*(a + b)
+    >>> poly.dorhs.factor()
+    a*x**2 + b*x + c*x**2 = x*(a*x**2 + b*x**2 + c)
+
+    ``poly.do.exp()`` or other sympy math functions will raise an error.
+
+    Rearranging an equation (simple example made complicated as illustration)
+    >>> p, V, n, R, T = var('p V n R T')
+    >>> eq1=Eqn(p*V,n*R*T)
+    >>> eq1
+    V*p = R*T*n
+    >>> eq2 =eq1/V
+    >>> eq2
+    p = R*T*n/V
+    >>> eq3 = eq2/R/T
+    >>> eq3
+    p/(R*T) = n/V
+    >>> eq4 = eq3*R/p
+    >>> eq4
+    1/T = R*n/(V*p)
+    >>> 1/eq4
+    T = V*p/(R*n)
+    >>> eq5 = 1/eq4 - T
+    >>> eq5
+    0 = -T + V*p/(R*n)
+
+    Substitution (#'s and units)
+    >>> L, atm, mol, K = var('L atm mol K', positive=True, real=True) # units
+    >>> eq2.subs({R:0.08206*L*atm/mol/K,T:273*K,n:1.00*mol,V:24.0*L})
+    p = 0.9334325*atm
+    >>> eq2.subs({R:0.08206*L*atm/mol/K,T:273*K,n:1.00*mol,V:24.0*L}).evalf(4)
+    p = 0.9334*atm
+
+    Combining equations (Math with equations: lhs with lhs and rhs with rhs)
+    >>> q = Eqn(a*c, b/c**2)
+    >>> q
+    a*c = b/c**2
+    >>> t
+    a = b/c
+    >>> q+t
+    a*c + a = b/c + b/c**2
+    >>> q/t
+    c = 1/c
+    >>> t**q
+    a**(a*c) = (b/c)**(b/c**2)
+
+    Utility operations
+    >>> t.reversed
+    b/c = a
+    >>> t.swap
+    b/c = a
+    >>> t.lhs
+    a
+    >>> t.rhs
+    b/c
+    >>> t.as_Boolean()
+    Eq(a, b/c)
+
+    `.check()` convenience method for `.as_Boolean().simplify()`
+    >>> from sympy import I, pi
+    >>> Equation(pi*(I+2), pi*I+2*pi).check()
+    True
+    >>> Eqn(a,a+1).check()
+    False
+
+    Differentiation
+    Differentiation is applied to both sides if the wrt variable appears on
+    both sides.
+    >>> q=Eqn(a*c, b/c**2)
+    >>> q
+    a*c = b/c**2
+    >>> diff(q,b)
+    Derivative(a*c, b) = c**(-2)
+    >>> diff(q,c)
+    a = -2*b/c**3
+    >>> diff(log(q),b)
+    Derivative(log(a*c), b) = 1/b
+    >>> diff(q,c,2)
+    Derivative(a, c) = 6*b/c**4
```

Comment:
This means that applying `doit()` on both sides of `diff(q, b)` will return `0 = c**(-2)`, right? I don't think that this is what we would expect.

Reply:
I agree that is not what we would expect. But the comments suggested there was no agreement on the behavior. The real problem is that the SymPy Diff operation makes the assumption that all symbols are independent of each other. Thus it is formally a partial derivative where all symbols except for the symbol of differentiation are assumed constant.

I am open to suggestions of how this should behave. We need some agreement. One idea is to have Eqn.doit(diff,x) raise a warning.